model_api/python/model_api/models/visual_prompting.py

```diff
@@ -0,0 +1,769 @@
+"""
+ Copyright (C) 2024 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from __future__ import annotations  # TODO: remove when Python3.9 support is dropped
+
+from collections import defaultdict
+from itertools import product
+from typing import Any, NamedTuple
+
+import cv2
+import numpy as np
+from model_api.models import SAMDecoder, SAMImageEncoder
+from model_api.models.utils import (
+    PredictedMask,
+    VisualPromptingResult,
+    ZSLVisualPromptingResult,
+)
+
+
+class VisualPromptingFeatures(NamedTuple):
+    feature_vectors: np.ndarray
+    used_indices: np.ndarray
+
+
+class Prompt(NamedTuple):
+    data: np.ndarray
+    label: int | np.ndarray
+
+
+class SAMVisualPrompter:
+    """
+    A wrapper that implements SAM Visual Prompter.
+
+    Segmentation results can be obtained by calling infer() method
+    with corresponding parameters.
+    """
+
+    def __init__(
+        self,
+        encoder_model: SAMImageEncoder,
+        decoder_model: SAMDecoder,
+    ):
+        self.encoder = encoder_model
+        self.decoder = decoder_model
+
+    def infer(
+        self,
+        image: np.ndarray,
+        boxes: list[Prompt] | None = None,
+        points: list[Prompt] | None = None,
+        polygons: list[Prompt] | None = None,
```

Comment:
Basic SAM inference doesn't need `polygons`.
If we already have `polygons`, we don't need to run `infer`.

Reply:
Deleted